### PR TITLE
tikv-ctl: compact whole cluster

### PR DESCRIPTION
**Pull request by UncP**
Compact store one by one is not very efficient, so introduce a new command that can compact the whole cluster in mutiple cfs at a time.

### CODE REVIEW

File: src/bin/tikv-ctl.rs [hunk 1]
```diff
@@ -1082,7 +1082,45 @@ fn main() {
...
+                )
+                .arg(
+                    Arg::with_name("db")
+                        .short("d")
+                        .takes_value(true)
+                        .default_value("kv")
+                        .help("kv or raft"),
+                )
+                .arg(
+                    Arg::with_name("cf")
```
Comment (huachaohuang) [inline src/bin/tikv-ctl.rs]: Can we take multiple values for this argument?

File: src/bin/tikv-ctl.rs [hunk 2]
```diff
@@ -1082,7 +1082,45 @@ fn main() {
                  .short("f").long("file")
                  .takes_value(true)
                  .help("meta file path"))
-        );
+        )
+        .subcommand(
+            SubCommand::with_name("compact-whole-cluster")
+                .about("compact whole cluster in a column family in a specified range")
```
Comment (huachaohuang) [inline src/bin/tikv-ctl.rs]: compact the whole cluster

File: src/bin/tikv-ctl.rs [hunk 3]
```diff
@@ -1297,3 +1344,31 @@ fn dump_snap_meta_file(path: &str) {
...
+    }
+
+    let pd_client = RpcClient::new(&cfg, Arc::clone(&mgr))
+        .unwrap_or_else(|e| perror_and_exit("RpcClient::new", e));
+
+    let stores = pd_client
+        .get_all_stores()
+        .unwrap_or_else(|e| perror_and_exit("Get all cluster stores from PD failed", e));
+
+    for e in stores {
```
Comment (huachaohuang) [inline src/bin/tikv-ctl.rs]: We should compact all stores concurrently.

File: src/bin/tikv-ctl.rs [hunk 4]
```diff
@@ -1100,17 +1138,26 @@ fn main() {
...
 
+    let mgr = new_security_mgr(&matches);
+
     if let Some(matches) = matches.subcommand_matches("dump-snap-meta") {
         let path = matches.value_of("file").unwrap();
         return dump_snap_meta_file(path);
+    } else if let Some(matches) = matches.subcommand_matches("compact-whole-cluster") {
+        let pd = matches.value_of("pd").unwrap();
+        let db = matches.value_of("db").unwrap();
+        let db_type = if db == "kv" { DBType::KV } else { DBType::RAFT };
```
Comment (Hoverbear) [inline src/bin/tikv-ctl.rs]: It would be good to throw an error if they user inputs something other than `kv` or `raft`.

File: src/bin/tikv-ctl.rs (hunk 2 above)
Comment (UncP) [reply src/bin/tikv-ctl.rs]: Change the command name or modify the help info?

File: src/bin/tikv-ctl.rs (hunk 4 above)
Comment (UncP) [reply src/bin/tikv-ctl.rs]: Yes, but this is fine since mostly we use tikv-ctl internally to debug, not those users.

File: src/bin/tikv-ctl.rs (hunk 1 above)
Comment (UncP) [reply src/bin/tikv-ctl.rs]: Done.

File: src/bin/tikv-ctl.rs [hunk 5]
```diff
@@ -1082,7 +1083,49 @@ fn main() {
                  .short("f").long("file")
                  .takes_value(true)
                  .help("meta file path"))
-        );
+        )
+        .subcommand(
+            SubCommand::with_name("compact-whole-cluster")
```
Comment (huachaohuang) [inline src/bin/tikv-ctl.rs]: "compact the whole cluster", both the name and the description.

File: src/bin/tikv-ctl.rs [hunk 6]
```diff
@@ -1082,7 +1083,49 @@ fn main() {
...
+                .about("compact whole cluster in a specified range in multiple column families")
+                .arg(
+                    Arg::with_name("pd")
+                        .short("p")
+                        .takes_value(true)
+                        .help("pd address"),
+                )
+                .arg(
+                    Arg::with_name("db")
```
Comment (huachaohuang) [inline src/bin/tikv-ctl.rs]: We can specify the possible values here.

File: src/bin/tikv-ctl.rs [hunk 7]
```diff
@@ -1082,7 +1083,49 @@ fn main() {
...
+                )
+                .arg(
+                    Arg::with_name("cf")
+                        .short("c")
+                        .takes_value(true)
+                        .multiple(true)
+                        .use_delimiter(true)
+                        .require_delimiter(true)
+                        .value_delimiter(",")
+                        .default_value(CF_DEFAULT)
```
Comment (huachaohuang) [inline src/bin/tikv-ctl.rs]: I think it's better to default to "default,write,lock".

File: src/bin/tikv-ctl.rs (hunk 6 above)
Comment (huachaohuang) [reply src/bin/tikv-ctl.rs]: https://kbknapp.github.io/clap-rs/clap/struct.Arg.html#method.possible_values

File: src/bin/tikv-ctl.rs [hunk 8]
```diff
@@ -1082,7 +1083,49 @@ fn main() {
...
+            SubCommand::with_name("compact-the-whole-cluster")
+                .about("compact the whole cluster in a specified range in one or more column families")
+                .arg(
+                    Arg::with_name("pd")
+                        .short("p")
+                        .takes_value(true)
+                        .help("pd address, eg. 127.0.0.1:2379"),
+                )
+                .arg(
+                    Arg::with_name("db")
```
Comment (huachaohuang) [inline src/bin/tikv-ctl.rs]: Please specify the possible values for this argument, like https://github.com/pingcap/tikv/blob/master/src/bin/tikv-server.rs#L300

File: src/bin/tikv-ctl.rs [hunk 9]
```diff
@@ -750,6 +758,14 @@ fn main() {
                 .takes_value(true)
                 .help("convert escaped key to hex key"),
         )
+        .arg(
+            Arg::with_name("pd")
```
Comment (huachaohuang) [inline src/bin/tikv-ctl.rs]: Seems other commands tend to keep the "pd" argument in the subcommand, I think we should do that too.
If you think we should move the "pd" argument to the upper level, you can discuss with others and maybe do that in another PR.

Comment (UncP) [reply src/bin/tikv-ctl.rs]: Yes, I tried that. Other commands put `pd` in subcommand  because they have one of the `db` or `host` argument. If we put `pd` in subcommand we must have one of the `db` or `host`, since they both have `.required(true)` field.

**Comment by siddontang** [comment]
please add a description for this PR.

**Comment by UncP** [comment]
Done.

File: src/bin/tikv-ctl.rs [hunk 10]
```diff
@@ -1297,3 +1360,44 @@ fn dump_snap_meta_file(path: &str) {
...
+        .unwrap_or_else(|e| perror_and_exit("Get all cluster stores from PD failed", e));
+
+    let mut handles = Vec::new();
+    for s in stores {
+        let mgr = Arc::clone(&mgr);
+        let addr = s.address.clone();
+        let (from, to) = (from.clone(), to.clone());
+        let cfs: Vec<String> = cfs.iter().map(|cf| cf.to_string().clone()).collect();
+        let h = thread::spawn(move || {
+            let debug_executor = new_debug_executor(None, None, Some(&addr), None, mgr);
```
Comment (hicqu) [inline src/bin/tikv-ctl.rs]: Please print every `addr`s.

File: src/bin/tikv-ctl.rs [hunk 11]
```diff
@@ -1082,7 +1098,45 @@ fn main() {
                  .short("f").long("file")
                  .takes_value(true)
                  .help("meta file path"))
-        );
+        )
+        .subcommand(
+            SubCommand::with_name("compact-the-whole-cluster")
```
Comment (hicqu) [inline src/bin/tikv-ctl.rs]: I think `compact-cluster` is enough.

File: src/bin/tikv-ctl.rs [hunk 12]
```diff
@@ -1082,7 +1098,45 @@ fn main() {
...
+                        .default_value(CF_DEFAULT)
+                        .possible_values(&["default", "lock", "write"])
+                        .help("column family names, for kv db, combine from default/lock/write; for raft db, can only be default"),
+                )
+                .arg(
+                    Arg::with_name("from")
+                        .short("f")
+                        .long("from")
+                        .takes_value(true)
+                        .help("set the start raw key, in escaped form"),
```
Comment (hicqu) [inline src/bin/tikv-ctl.rs]: Now we have a `raw_key_hint` to unify the help message.

**Comment by hicqu** [comment]
rest LGTM.

File: src/bin/tikv-ctl.rs (hunk 10 above)
Comment (UncP) [reply src/bin/tikv-ctl.rs]: It's better to print them when compaction is done since we are in multi-thread env.

**Comment by UncP** [comment]
@hicqu @huachaohuang

File: src/bin/tikv-ctl.rs [hunk 13]
```diff
@@ -561,15 +577,29 @@ impl DebugExecutor for DebugClient {
         ) as Box<Stream<Item = (Vec<u8>, MvccInfo), Error = String>>
     }
 
-    fn do_compact(&self, db: DBType, cf: &str, from: Vec<u8>, to: Vec<u8>) {
+    fn do_compaction(
+        &self,
+        address: Option<&str>,
```
Comment (huachaohuang) [inline src/bin/tikv-ctl.rs]: Why not log it outside so that you don't need to pass the address here.

File: src/bin/tikv-ctl.rs [hunk 14]
```diff
@@ -1200,17 +1281,26 @@ fn main() {
         _ => unreachable!(),
     };
 
+    let mgr = new_security_mgr(&matches);
+
     if let Some(matches) = matches.subcommand_matches("dump-snap-meta") {
         let path = matches.value_of("file").unwrap();
         return dump_snap_meta_file(path);
+    } else if let Some(sub_cmd) = matches.subcommand_matches("compact-the-whole-cluster") {
```
Comment (huachaohuang) [inline src/bin/tikv-ctl.rs]: s/compact-the-whole-cluster/compact-cluster

File: src/bin/tikv-ctl.rs [hunk 15]
```diff
@@ -1416,3 +1496,44 @@ fn dump_snap_meta_file(path: &str) {
...
+    let mut handles = Vec::new();
+    for s in stores {
+        let mgr = Arc::clone(&mgr);
+        let addr = s.address.clone();
+        let (from, to) = (from.clone(), to.clone());
+        let cfs: Vec<String> = cfs.iter().map(|cf| cf.to_string().clone()).collect();
+        let h = thread::spawn(move || {
+            let debug_executor = new_debug_executor(None, None, Some(&addr), None, mgr);
+            for cf in cfs {
+                debug_executor.compact(Some(&addr), db_type, cf.as_str(), from.clone(), to.clone());
```
Comment (huachaohuang) [inline src/bin/tikv-ctl.rs]: I mean, if you log here, you don't need to pass the address in.

Comment (UncP) [reply src/bin/tikv-ctl.rs]: If log here, the `println` still can't be removed since there are `compact` and `compact_region` commands. It's better to use one `println` instead of two.

**Comment by hicqu** [comment]
LGTM.
